runtime/compiler/optimizer/HotFieldMarking.cpp

```diff
@@ -0,0 +1,173 @@
+         {
+         block = node->getBlock();
+         }
+      else if ((node->getOpCode().isLoadIndirect() || node->getOpCode().isStoreIndirect())
+               && node->getOpCode().hasSymbolReference()
+               && !node->getSymbolReference()->isUnresolved()
+               && node->getSymbolReference()->getSymbol()->isShadow()
+               && !node->isInternalPointer()
+               && !node->getOpCode().isArrayLength()
+               && node->getSymbolReference()->getSymbol()->isCollectedReference()
+              )
+         {
+         TR::SymbolReference *symRef = node->getSymbolReference();
+         if (symRef->getCPIndex() >= 0 && !symRef->getSymbol()->isArrayShadowSymbol())
+            {
+            auto itr = stats.find(symRef->getSymbol());
+            if (itr != stats.end())
+               {
+               if(TR::Options::getReductionAlgorithm(TR_HotFieldReductionAlgorithmSum))
+                  {
+                  itr->second->_score = getReducedFrequencySum(itr->second->_score, block->getFrequency());
+                  } 
+               else
+                  {
+                  itr->second->_score = getReducedFrequencyAverage(itr->second->_score, itr->second->_count, block->getFrequency());
+                  itr->second->_count += 1;
+                  }
+               continue;
+               }
+
+            TR::ResolvedMethodSymbol *rms = comp()->getOwningMethodSymbol(symRef->getOwningMethodIndex());
+            TR_ResolvedMethod *method = rms->getResolvedMethod();
+
+            int32_t fieldNameLength = 0;
+            char *fieldName = method->fieldNameChars(symRef->getCPIndex(), fieldNameLength);
+            int32_t fieldSigLength = 0;
+            char *fieldSig = method->fieldSignatureChars(symRef->getCPIndex(), fieldSigLength);
+            bool isStatic = false;
+            TR_OpaqueClassBlock *containingClass = static_cast<TR_ResolvedJ9Method*>(method)->definingClassFromCPFieldRef(comp(), symRef->getCPIndex(), isStatic);
+            if (isStatic)
+               continue;
+
+            stats[symRef->getSymbol()] = new (trStackMemory()) SymStats(1, block->getFrequency(), fieldNameLength, fieldName, fieldSigLength, fieldSig, containingClass);
+            }
+         }
+      }
+
+   for (auto itr = stats.begin(), end = stats.end(); itr != end; ++itr)
+      {
+      if (itr->second->_score >= TR::Options::_hotFieldThreshold)
+         {
+         int32_t fieldOffset = (comp()->fej9()->getInstanceFieldOffset(itr->second->_clazz, itr->second->_fieldName, itr->second->_fieldNameLength, itr->second->_fieldSig, itr->second->_fieldSigLength) + TR::Compiler->om.objectHeaderSizeInBytes()) / TR::Compiler->om.sizeofReferenceField();
+            
+         if (!comp()->fej9()->isAnonymousClass(itr->second->_clazz))
+            {
+            comp()->fej9()->reportHotField(getUtilization(), TR::Compiler->cls.convertClassOffsetToClassPtr(itr->second->_clazz), fieldOffset, itr->second->_score);
```

Comment:
usually the side-effects of an opt pass are guarded by performTransformation calls facilitates binary searching of problematic optimizations - we might want to consider that here to allow for filtering of the field data being reported.